include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -230,15 +230,27 @@ def ReadLValueOp : MooreOp<"read_lvalue", [SameOperandsAndResultType]> {
   }];
 }
 
+def ReadOp : MooreOp<"read"> {
+  let summary = "Like llvm.load";
+  let description = [{
+    //TODO:
+  }];
+  let arguments = (ins Arg<RefType, "", [MemRead]>:$input);
+  let results = (outs UnpackedType:$result);
+  let assemblyFormat = [{
+    $input attr-dict `:` type($input) `->` type($result)
+  }];
+}
+
 //===----------------------------------------------------------------------===//
 // Assignments
 //===----------------------------------------------------------------------===//
 
 class AssignOpBase<string mnemonic, list<Trait> traits = []> :
-    MooreOp<mnemonic, traits # [SameTypeOperands]> {
-  let arguments = (ins UnpackedType:$dst, UnpackedType:$src);
+    MooreOp<mnemonic, traits # []> {
+  let arguments = (ins RefType:$dst, UnpackedType:$src);
   let assemblyFormat = [{
-    $dst `,` $src attr-dict `:` type($src)
+    $dst `,` $src attr-dict `:` type($dst) `,` type($src)
```

Comment:
You can use `TypesMatchWith` constraints here to encode that `dst` has to have type `ref<T>` and src type `T`. Similar to the suggestions above. Check out the `sv.assign` op in the SV dialect which does a similar thing. That also allows you to drop `type($dst)`.